modules/siddhi-extensions/input-transports/jms-input-transport/src/main/java/org/wso2/siddhi/extension/input/transport/jms/JMSWorkerThread.java

```diff
@@ -45,22 +45,27 @@ public JMSWorkerThread(CarbonMessage cMessage, CarbonCallback cCallback, SourceE
 
     @Override
     public void run() {
-        if (carbonMessage instanceof TextCarbonMessage) {
-            String event = ((TextCarbonMessage) carbonMessage).getText();
-            sourceEventListener.onEvent(event);
-        } else if (carbonMessage instanceof MapCarbonMessage) {
-            Map<String, String> event = new HashMap<>();
-            Enumeration<String> mapNames = ((MapCarbonMessage) carbonMessage).getMapNames();
-            while (mapNames.hasMoreElements()) {
-                String key = mapNames.nextElement();
-                event.put(key, ((MapCarbonMessage) carbonMessage).getValue(key));
+        try {
+            if (carbonMessage instanceof TextCarbonMessage) {
+                String event = ((TextCarbonMessage) carbonMessage).getText();
+                sourceEventListener.onEvent(event);
+            } else if (carbonMessage instanceof MapCarbonMessage) {
+                Map<String, String> event = new HashMap<>();
+                Enumeration<String> mapNames = ((MapCarbonMessage) carbonMessage).getMapNames();
+                while (mapNames.hasMoreElements()) {
+                    String key = mapNames.nextElement();
+                    event.put(key, ((MapCarbonMessage) carbonMessage).getValue(key));
+                }
+                sourceEventListener.onEvent(event);
+            } else {
+                throw new JMSInputAdaptorRuntimeException("The message type of the JMS message is not supported!");
             }
-            sourceEventListener.onEvent(event);
-        } else {
-            throw new JMSInputAdaptorRuntimeException("The message type of the JMS message is not supported!");
-        }
-        if (carbonCallback != null) {
-            carbonCallback.done(carbonMessage);
+            // ACK only if the event is processed i.e: no exceptions thrown from the onEvent method.
+            if (carbonCallback != null) {
+                carbonCallback.done(carbonMessage);
+            }
+        } catch (RuntimeException e) {
+            throw new JMSInputAdaptorRuntimeException("Failed to process JMS message.", e);
```

Comment:
Not exactly. This exception is thrown if the message cannot be processed and we don't want to ack the JMS broker that the message was processed. So we will get the same message again.